File: ios/PulseWatchKit/Sources/PulseWatchAPI/APIClient.swift
import Foundation
import PulseWatchModels
import PulseWatchVault

public enum APIError: Error, LocalizedError {
    case http(status: Int, body: String)
    case decode(Error)
    case transport(Error)
    case notAuthenticated

    public var errorDescription: String? {
        switch self {
        case .http(let s, let b): return "HTTP \(s): \(b)"
        case .decode(let e): return "Decode: \(e.localizedDescription)"
        case .transport(let e): return e.localizedDescription
        case .notAuthenticated: return "Not signed in"
        }
    }
}

public final class APIClient: @unchecked Sendable {
    public let baseURL: URL
    public let session: SessionStore
    private let urlSession: URLSession
    private let decoder: JSONDecoder
    private let encoder: JSONEncoder

    public init(baseURL: URL, session: SessionStore = SessionStore(), urlSession: URLSession = .shared) {
        self.baseURL = baseURL
        self.session = session
        self.urlSession = urlSession
        let d = JSONDecoder(); d.keyDecodingStrategy = .useDefaultKeys
        self.decoder = d
        let e = JSONEncoder(); e.keyEncodingStrategy = .useDefaultKeys
        self.encoder = e
    }

    // MARK: auth

    public func enrollDevice(
        platform: DevicePlatform,
        publicKey: Data,
        pairingCode: String? = nil,
        deviceName: String?
    ) async throws -> Session {
        let body = DeviceEnrollRequest(
            platform: platform,
            public_key: publicKey.base64EncodedString(),
            pairing_code: pairingCode,
            device_name: deviceName
        )
        return try await send("/v1/auth/devices", method: "POST", body: body, authed: false)
    }

    public func createPairing() async throws -> Pairing {
        try await send("/v1/auth/pairings", method: "POST", body: EmptyBody())
    }

    // MARK: wrapping keys

    public func currentWrappingKey() async throws -> WrappingKey {
        try await send("/v1/wrapping-keys/current", method: "GET")
    }

    // MARK: accounts

    public func listAccounts() async throws -> [Account] {
        try await send("/v1/accounts", method: "GET")
    }

    public func createAccount(_ create: AccountCreate) async throws -> Account {
        try await send("/v1/accounts", method: "POST", body: create)
    }

    public func deleteAccount(id: String) async throws {
        let _: EmptyResponse = try await send("/v1/accounts/\(id)", method: "DELETE", expectEmpty: true)
    }

    public func refreshAccount(id: String) async throws {
        let _: EmptyResponse = try await send("/v1/accounts/\(id)/refresh", method: "POST", expectEmpty: true)
    }

    // MARK: usage

    public func usageSummary(window: UsageWindow = .day) async throws -> UsageSummary {
        try await send("/v1/usage/summary?window=\(window.rawValue)", method: "GET")
    }

    public func usageSeries(
        accountId: String? = nil,
        from: Date,
        to: Date,
        bucket: UsageBucket = .day,
        groupBy: String = "none"
    ) async throws -> UsageSeries {
        let iso = ISO8601DateFormatter()
        var items = [
            URLQueryItem(name: "from", value: iso.string(from: from)),
            URLQueryItem(name: "to", value: iso.string(from: to)),
            URLQueryItem(name: "bucket", value: bucket.rawValue),
            URLQueryItem(name: "group_by", value: groupBy)
        ]
        if let accountId { items.append(URLQueryItem(name: "account_id", value: accountId)) }
        var comps = URLComponents()
        comps.queryItems = items
        let q = comps.percentEncodedQuery ?? ""
        return try await send("/v1/usage/series?\(q)", method: "GET")
    }

    // MARK: alerts

    public func listThresholds() async throws -> [AlertThreshold] {
        try await send("/v1/alerts/thresholds", method: "GET")
    }

    public func putThresholds(_ items: [AlertThresholdInput]) async throws -> [AlertThreshold] {
        try await send("/v1/alerts/thresholds", method: "PUT", body: items)
    }

    // MARK: push

    public func putPushToken(_ input: PushTokenInput) async throws {
        let _: EmptyResponse = try await send(
            "/v1/devices/push-token", method: "PUT", body: input, expectEmpty: true
        )
    }

    public func deletePushToken() async throws {
        let _: EmptyResponse = try await send(
            "/v1/devices/push-token", method: "DELETE", expectEmpty: true
        )
    }

    // MARK: low-level

    private struct EmptyResponse: Decodable {}
    private struct EmptyBody: Encodable {}

    private func send<R: Decodable>(
        _ path: String,
        method: String,
        body: Encodable? = nil,
        authed: Bool = true,
        expectEmpty: Bool = false
    ) async throws -> R {
        let trimmedBase = baseURL.absoluteString.hasSuffix("/")
            ? String(baseURL.absoluteString.dropLast()) : baseURL.absoluteString
        let prefix = path.hasPrefix("/") ? "" : "/"
        guard let url = URL(string: trimmedBase + prefix + path) else {
            throw APIError.http(status: -1, body: "bad url")
        }
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "accept")

        if authed {
            guard let s = session.load() else { throw APIError.notAuthenticated }
            req.setValue("Bearer \(s.token)", forHTTPHeaderField: "authorization")
        }

        if let body {
            req.setValue("application/json", forHTTPHeaderField: "content-type")
            req.httpBody = try encoder.encode(AnyEncodable(body))
        }

        let (data, response): (Data, URLResponse)
        do {
            (data, response) = try await urlSession.data(for: req)
        } catch {
            throw APIError.transport(error)
        }
        guard let http = response as? HTTPURLResponse else {
            throw APIError.http(status: -1, body: "no response")
        }
        guard (200..<300).contains(http.statusCode) else {
            throw APIError.http(status: http.statusCode, body: String(data: data, encoding: .utf8) ?? "")
        }
        if expectEmpty {
            return try decoder.decode(R.self, from: Data("{}".utf8))
        }
        do {
            return try decoder.decode(R.self, from: data)
        } catch {
            throw APIError.decode(error)
        }
    }
}

// Erases Encodable so the `body` parameter can be heterogeneous.
private struct AnyEncodable: Encodable {
    let value: Encodable
    init(_ v: Encodable) { self.value = v }
    func encode(to encoder: Encoder) throws { try value.encode(to: encoder) }
}

File: android/core/src/main/java/app/cap/core/api/ApiClient.kt
package app.cap.core.api

import app.cap.core.vault.SessionStore
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import java.io.IOException
import java.util.concurrent.TimeUnit

class ApiException(val status: Int, body: String) : IOException("HTTP $status: $body")

class ApiClient(
    val baseUrl: String,
    val sessionStore: SessionStore,
    private val http: OkHttpClient = defaultClient(),
) {
    @PublishedApi internal val json: Json = Json { ignoreUnknownKeys = true; encodeDefaults = false }
    @PublishedApi internal val jsonMedia = "application/json".toMediaType()

    // Auth ------------------------------------------------------------------

    suspend fun enrollDevice(req: DeviceEnrollRequest): Session =
        post("/v1/auth/devices", req, authed = false)

    suspend fun createPairing(): Pairing = post<Unit?, Pairing>("/v1/auth/pairings", body = null)

    // Wrapping keys ---------------------------------------------------------

    suspend fun currentWrappingKey(): WrappingKey = get("/v1/wrapping-keys/current")

    // Accounts --------------------------------------------------------------

    suspend fun listAccounts(): List<Account> = get("/v1/accounts")
    suspend fun createAccount(req: AccountCreate): Account = post("/v1/accounts", req)
    suspend fun deleteAccount(id: String) {
        empty("/v1/accounts/$id", "DELETE")
    }

    suspend fun refreshAccount(id: String) {
        empty("/v1/accounts/$id/refresh", "POST")
    }

    // Usage -----------------------------------------------------------------

    suspend fun usageSummary(window: UsageWindow = UsageWindow.DAY): UsageSummary {
        val w = window.name.lowercase()
        return get("/v1/usage/summary?window=$w")
    }

    suspend fun usageSeries(
        from: String,
        to: String,
        bucket: UsageBucket = UsageBucket.DAY,
        accountId: String? = null,
        groupBy: String = "none",
    ): UsageSeries {
        val q = buildString {
            append("from=").append(from)
            append("&to=").append(to)
            append("&bucket=").append(bucket.name.lowercase())
            append("&group_by=").append(groupBy)
            if (accountId != null) append("&account_id=").append(accountId)
        }
        return get("/v1/usage/series?$q")
    }

    // Alerts ----------------------------------------------------------------

    suspend fun listThresholds(): List<AlertThreshold> = get("/v1/alerts/thresholds")
    suspend fun putThresholds(items: List<AlertThresholdInput>): List<AlertThreshold> =
        put("/v1/alerts/thresholds", items)

    // Push ------------------------------------------------------------------

    suspend fun putPushToken(input: PushTokenInput) {
        val payload = json.encodeToString(serializer<PushTokenInput>(), input)
        execute("/v1/devices/push-token", "PUT", payload, authed = true).close()
    }

    suspend fun deletePushToken() {
        empty("/v1/devices/push-token", "DELETE")
    }

    // Internals -------------------------------------------------------------

    private suspend inline fun <reified Res> get(path: String): Res =
        execute(path, "GET", body = null, authed = true).use { readBody(it) }

    private suspend inline fun <reified Req, reified Res> post(
        path: String, body: Req?, authed: Boolean = true,
    ): Res {
        val payload = body?.let { json.encodeToString(serializer(), it) }
        return execute(path, "POST", payload, authed).use { readBody(it) }
    }

    private suspend inline fun <reified Req, reified Res> put(path: String, body: Req): Res {
        val payload = json.encodeToString(serializer(), body)
        return execute(path, "PUT", payload, authed = true).use { readBody(it) }
    }

    private suspend fun empty(path: String, method: String) {
        execute(path, method, body = null, authed = true).close()
    }

    @PublishedApi
    internal inline fun <reified T> readBody(response: Response): T {
        val text = response.body!!.string()
        return json.decodeFromString(serializer(), text)
    }

    @PublishedApi
    internal suspend fun execute(
        path: String,
        method: String,
        body: String?,
        authed: Boolean,
    ): Response = withContext(Dispatchers.IO) {
        val url = (baseUrl.trimEnd('/') + path).toHttpUrl()
        val rb = body?.toRequestBody(jsonMedia)
            ?: if (method in setOf("POST", "PUT", "DELETE")) "".toRequestBody(jsonMedia) else null
        val builder = Request.Builder().url(url).method(method, rb).header("Accept", "application/json")
        if (authed) {
            val s = sessionStore.load() ?: throw ApiException(401, "no session")
            builder.header("Authorization", "Bearer ${s.token}")
        }
        val response = http.newCall(builder.build()).execute()
        if (!response.isSuccessful) {
            val text = response.body?.string().orEmpty()
            response.close()
            throw ApiException(response.code, text)
        }
        response
    }

    companion object {
        private fun defaultClient(): OkHttpClient = OkHttpClient.Builder()
            .connectTimeout(15, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .callTimeout(60, TimeUnit.SECONDS)
            .build()
    }
}

// re-export so callers don't need to depend on kotlinx.serialization directly
@PublishedApi internal inline fun <reified T> serializer(): kotlinx.serialization.KSerializer<T> =
    kotlinx.serialization.serializer()
